File: features/fintech/instant_loan/src/main/java/com/tokopedia/instantloan/di/module/InstantLoanModule.java
package com.tokopedia.instantloan.di.module;

import android.content.Context;

import com.tokopedia.abstraction.common.di.qualifier.ApplicationContext;
import com.tokopedia.abstraction.common.di.scope.ApplicationScope;
import com.tokopedia.abstraction.common.network.interceptor.HeaderErrorResponseInterceptor;
import com.tokopedia.instantloan.di.scope.InstantLoanScope;
import com.tokopedia.instantloan.domain.interactor.GetBannersUserCase;
import com.tokopedia.instantloan.domain.interactor.GetLoanProfileStatusUseCase;
import com.tokopedia.instantloan.domain.interactor.PostPhoneDataUseCase;
import com.tokopedia.instantloan.network.InstantLoanAuthInterceptor;
import com.tokopedia.user.session.UserSession;

import dagger.Module;
import dagger.Provides;
import okhttp3.Interceptor;
import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava.RxJavaCallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;

import static com.tokopedia.instantloan.network.InstantLoanUrl.BaseUrl.WEB_DOMAIN;

/**
 * Created by lavekush on 19/03/18.
 */
@InstantLoanScope
@Module
public class InstantLoanModule {

    @Provides
    public GetLoanProfileStatusUseCase provideGetLoanProfileStatusUseCase(InstantLoanAuthInterceptor instantLoanAuthInterceptor, @ApplicationContext Context context) {
        return new GetLoanProfileStatusUseCase(instantLoanAuthInterceptor, context);
    }

    @Provides
    public GetBannersUserCase provideGetBannersUseCase(InstantLoanAuthInterceptor instantLoanAuthInterceptor, @ApplicationContext Context context) {
        return new GetBannersUserCase(instantLoanAuthInterceptor, context);
    }

    @Provides
    public PostPhoneDataUseCase providePostPhoneDataUseCase(InstantLoanAuthInterceptor instantLoanAuthInterceptor, @ApplicationContext Context context) {
        return new PostPhoneDataUseCase(instantLoanAuthInterceptor, context);
    }

    @InstantLoanScope
    @Provides
    public UserSession provideUserSession(@ApplicationContext Context context) {
        return new UserSession(context);
    }
}

File: features/fintech/instant_loan/src/main/java/com/tokopedia/instantloan/di/module/InstantLoanModule.kt
package com.tokopedia.instantloan.di.module

import android.content.Context
import com.tokopedia.abstraction.common.di.qualifier.ApplicationContext
import com.tokopedia.instantloan.di.scope.InstantLoanScope
import com.tokopedia.instantloan.domain.interactor.GetFilterDataUseCase
import com.tokopedia.instantloan.domain.interactor.GetLendingDataUseCase
import com.tokopedia.instantloan.domain.interactor.GetLoanProfileStatusUseCase
import com.tokopedia.instantloan.domain.interactor.PostPhoneDataUseCase
import com.tokopedia.instantloan.network.InstantLoanAuthInterceptor
import com.tokopedia.network.NetworkRouter
import com.tokopedia.user.session.UserSession
import com.tokopedia.user.session.UserSessionInterface
import dagger.Module
import dagger.Provides

@InstantLoanScope
@Module
class InstantLoanModule {

    @InstantLoanScope
    @Provides
    fun provideUserSessionInterface(@ApplicationContext context: Context): UserSessionInterface {
        return UserSession(context)
    }


    @Provides
    fun provideNetworkRouter(@ApplicationContext context: Context): NetworkRouter {
        return context as NetworkRouter
    }

    @Provides
    fun provideGetLoanProfileStatusUseCase(instantLoanAuthInterceptor: InstantLoanAuthInterceptor, @ApplicationContext context: Context): GetLoanProfileStatusUseCase {
        return GetLoanProfileStatusUseCase(instantLoanAuthInterceptor, context)
    }

    @Provides
    fun provideGetLendingDataUseCase(@ApplicationContext context: Context) : GetLendingDataUseCase {
        return GetLendingDataUseCase(context)
    }

    @Provides
    fun provideGetFilterDataUseCase(@ApplicationContext context: Context) : GetFilterDataUseCase {
        return GetFilterDataUseCase(context)
    }

    @Provides
    fun providePostPhoneDataUseCase(instantLoanAuthInterceptor: InstantLoanAuthInterceptor, @ApplicationContext context: Context): PostPhoneDataUseCase {
        return PostPhoneDataUseCase(instantLoanAuthInterceptor, context)
    }

    @InstantLoanScope
    @Provides
    fun provideUserSession(@ApplicationContext context: Context): UserSession {
        return UserSession(context)
    }
}

